### PR TITLE
[#11][Backend] As a user, I can sign out

### PR DESCRIPTION
**Pull request by hoangnguyen92dn**
Close https://github.com/hoangnguyen92dn/survey-flutter-ic/issues/11

## What happened 👀

Call `POST /api/v1/oauth/revoke` to sign out with the following body:

```
{
  "token": "{{user_access_token}}",
  "client_id": "{{client_id}}",
  "client_secret": "{{client_secret}}"
}
```

## Insight 📝

- Add API request to handle sign out
- Add UnitTest

## Proof Of Work 📹

```
I/flutter (22892): *** Request ***
I/flutter (22892): uri: https://survey-api.nimblehq.co/api/v1/oauth/revoke
I/flutter (22892): method: POST
I/flutter (22892): responseType: ResponseType.json
I/flutter (22892): followRedirects: true
I/flutter (22892): persistentConnection: true
I/flutter (22892): connectTimeout: 0:00:03.000000
I/flutter (22892): sendTimeout: null
I/flutter (22892): receiveTimeout: 0:00:05.000000
I/flutter (22892): receiveDataWhenStatusError: true
I/flutter (22892): extra: {}
I/flutter (22892): headers:
I/flutter (22892):  Content-Type: application/json; charset=utf-8
I/flutter (22892):  Authorization: Bearer h9BXhSKIGnZGx1OHtFQL14QTifEE77Ey2r49wIX-TGA
I/flutter (22892): data:
I/flutter (22892): {token: h9BXhSKIGnZGx1OHtFQL14QTifEE77Ey2r49wIX-TGA, client_id: 6GbE8dhoz519l2N_F99StqoOs6Tcmm1rXgda4q__rIw, client_secret: _ayfIm7BeUAhx2W1OUqi20fwO3uNxfo1QstyKlFCgHw}
I/flutter (22892): 
I/flutter (22892): *** Response ***
I/flutter (22892): uri: https://survey-api.nimblehq.co/api/v1/oauth/revoke
I/flutter (22892): statusCode: 200
I/flutter (22892): headers:
I/flutter (22892):  connection: keep-alive
I/flutter (22892):  cache-control: max-age=0, private, must-revalidate
I/flutter (22892):  date: Wed, 26 Apr 2023 04:22:15 GMT
I/flutter (22892):  vary: Accept-Encoding, Origin
I/flutter (22892):  content-encoding: gzip
I/flutter (22892):  referrer-policy: strict-origin-when-cross-origin
I/flutter (22892):  report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=xeVTSEqZ3t3KGXSB%2FerXApjwIUELb9rHdFW5eauUXtWVKIPynhUhn0RppDRihg0x9IUnU9PgmFKcLIkWMMr8QzO3dUs5HWDb%2BC7u9tJHN32anHjxAgmz%2BoaqjgUJwLCfPZ7%2Bh7S6X8RL"}],"group":"cf-nel","max_age":604800}
I/flutter (22892):  cf-cache-status: DYNAMIC
I/flutter (22892):  x-permitted-cross-domain-policies: none
I/flutter (22892):  content-type: application/json; charset=utf-8
I/flutter (22892):  x-xss-protection: 0
I/flutter (22892):  server: cloudflare
I/flutter (22892):  x-request-id: e8015dab-5559-4474-ab73-c1711f78d085
I/flutter (22892):  alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400
I/flutter (22892):  content-length: 28
I/flutter (22892):  nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
I/flutter (22892):  x-download-options: noopen
I/flutter (22892):  cf-ray: 7bdc0b2499e80445-HKG
I/flutter (22892):  x-runtime: 0.010318
I/flutter (22892):  etag: W/"b9ec4c2a7a080b1916331a8d944a2402"
I/flutter (22892):  via: 1.1 vegur
I/flutter (22892):  x-frame-options: SAMEORIGIN
I/flutter (22892):  x-content-type-options: nosniff
I/flutter (22892): Response Text:
I/flutter (22892): {}
I/flutter (22892): 
```

### CODE REVIEW

**Comment by manh-t** [comment]
@hoangnguyen92dn please rebase with the newest changes on develop. There are some unrelated commits in your PR 😄 
image

**Comment by hoangnguyen92dn** [comment]
@manh-t Rebased 🚀

File: lib/usecase/sign_out_use_case.dart [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  @override
+  Future<Result<void>> call() async {
+    final token = await _persistence.accessToken ?? '';
+    return _repository
+        .signOut(token: token)
+        // ignore: unnecessary_cast
+        .then((token) => Success(null) as Result<void>)
+        .onError<NetworkExceptions>(
+            (exception, stackTrace) => Failed(UseCaseException(exception)));
+  }
```
Comment (manh-t) [inline lib/usecase/sign_out_use_case.dart:26]: @hoangnguyen92dn I think we should clear the token in `persistence` when calling `signOut`

Comment (hoangnguyen92dn) [reply lib/usecase/sign_out_use_case.dart:26]: @manh-t Indeed. Please allow me to add the clear tokens on the integration PR

https://github.com/hoangnguyen92dn/survey-flutter-ic/pull/76/files#diff-b5da62b96f31add08b94efede14c43dbe0c9c13eae063c3f54d51f864603b054R23

**Review by manh-t — APPROVED** [review]

**Pull request merged**